content/2021-06-11-make-your-life-easier.md

```diff
@@ -0,0 +1,156 @@
+operate on them gets noticeably harder as you work your way down the list. If
+you can constrain your requirements to allow you to use structures nearer the
+top of that list you will have fewer implementation bugs.
+
+Sometimes you can't avoid a cyclic graph, but by remembering that it's harder
+than the others you should stop and think "do I need more tests for this?"
+
+# Default to immutable data
+
+Cache invalidation is hard. If your data is immutable you rarely have to solve
+that problem. Immutable data is also easier to archive and/or store in cheap
+blob stores like S3 or in static files on a filesystem.
+
+It can be very hard to make your data truly immutable. For example, the [GDPR
+"right to be forgotten"](https://gdpr-info.eu/art-17-gdpr/) means that
+anything containing [personal data](https://gdpr-info.eu/art-4-gdpr/) must be
+deletable. So just like with graph data structures there's a kind of hierarchy
+of mutability of data structures:
+
+- Truly immutable. The data is created and is never destroyed.
+- Can be atomicly created or destroyed but never modified.
+- Can have portions appended or deleted, but previous data is never modified.
+- Some portions of the data structure are immutable, others can be deleted or
+  mutated in place.
+- All parts of the data structure are mutable.
+
+Stay as far up that list as you can and you will avoid many implementation
+problems.
+
+# Choose simpler programming models
+
+There are lots of techniques to improve the performance of software. Use none
+of them until you are forced to. Concurrency is hard to get right.
+Event-driven systems can be difficult to reason about. Manual memory
+management is a notorious source of bugs. Functions without side-effects are
+easier to test than functions with them.
+
+Start with single-threaded, single process code in the highest-level language
+you can use and iterate from there. Use pure functions when you can, but don't
+be a zealot about it. The ["Functional Core, Imperative
+Shell"](https://www.destroyallsoftware.com/screencasts/catalog/functional-core-imperative-shell)
+pattern can be very useful.
+
+# Worry about all your public interfaces
+
+[Hyrum's law](https://www.hyrumslaw.com/) teaches us that anything your
+customer can perceive about your software is part of its public interface. If
+you change it, [someone will complain](https://xkcd.com/1172/).
+
+When you're designing new software, divide your problem space into features
+and behaviours that are visible and invisible to customers. Put the bulk of
+your effort into the visible parts, everything else is fixable later.
+
+If you don't want your customer to depend on something, make it impossible to
+depend upon. For example, if you have a function that returns a list of items
+and for some reason you don't want the ordering to be defined then you may
+need to deliberately shuffle the items to prevent people from depending on the
+order.
+
+# You probably can't avoid distributed systems
```

Comment:
This section is juicy and feels like its own post relative to the other sections

Reply:
Good to know. I might come back to it later and flesh it out as another post.